### PR TITLE
attention mask & bias

### PR DESCRIPTION
**Pull request by robotcator**
Add attention mask & bias.

Currently, because of the ldg instruction problem, we only support an even number of sequence lengths.

### CODE REVIEW

File: csrc/flash_attn/fmha_api.cpp [hunk 1]
```diff
@@ -385,6 +450,41 @@ mha_bwd(const at::Tensor &dout,  // total_q x num_heads, x head_size
...
+        mask_seq_mod_size = mask_sizes[2];
+        TORCH_CHECK(mask_sizes[1] == 1 || mask_sizes[1] == num_heads);
+        TORCH_CHECK(mask_sizes[2] == 1 || mask_sizes[2] == max_seqlen_q_);
+    }
+
+    auto opts = q.options();
+    at::Tensor ds;
+    if (attn_bias.has_value()) {
+        ds = torch::empty({batch_size, num_heads, max_seqlen_q_, max_seqlen_k_}, opts.dtype(q_dtype));
+        ds.zero_();
```
Comment (guolinke) [inline csrc/flash_attn/fmha_api.cpp:484]: ds?

File: csrc/flash_attn/src/fmha_fprop_kernel_1xN.h [hunk 2]
```diff
@@ -392,7 +418,32 @@ inline __device__ void device_1xN_(const Params &params, const int bidb, const i
...
+            fmha::clear(frag_mask);
+            gmem_mask.template load<Frag_mask, elem_type>(frag_mask);
+            gmem_mask.move();
+
+            // Apply the attn mask.
+            softmax.apply_attn_mask(frag_mask, l, loop_step_idx);
+        }
+
+        if constexpr (has_bias) {
+        // if (!(params.attn_bias_ptr == nullptr)) {
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_fprop_kernel_1xN.h]: remove these unuse codes?

File: csrc/flash_attn/src/fmha_dgrad_fp16_kernel_loop.sm80.cu [hunk 3]
```diff
@@ -28,28 +28,103 @@ void run_fmha_dgrad_fp16_sm80_loop_(const FMHA_dgrad_params &params, cudaStream_
...
+                dim3 grid(params.b, params.h);
+                kernel<<<grid, Kernel_traits::THREADS, smem_size_dq_dk_dv, stream>>>(params);
+                FMHA_CHECK_CUDA(cudaPeekAtLastError());
+            });
         }
-        if( smem_size_dq_dk_dv >= 48 * 1024 ) {
-            FMHA_CHECK_CUDA(cudaFuncSetAttribute(
-                kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_size_dq_dk_dv));
+    }else{
+        if (has_bias) {
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_dgrad_fp16_kernel_loop.sm80.cu]: constexpr

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h [hunk 4]
```diff
@@ -197,6 +211,15 @@ inline __device__ void compute_dq_dk_dv_1xN_one_iter(const Params &params, Prng
     gmem_softmax_lse.move(begin);
     gmem_softmax_d.move(begin);
 
+    if constexpr (has_attn) {
+        gmem_mask.move(begin);
+    }
+
+    if constexpr (has_attn) {
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h]: has_attn or has_bias?

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h [hunk 5]
```diff
@@ -148,6 +148,20 @@ inline __device__ void compute_dq_dk_dv_1xN_one_iter(const Params &params, Prng
...
+    // bool has_attn = !(params.attn_mask_ptr == nullptr);
+    // Allocate the global memory tile loader for mask.
+    using Gmem_tile_mask = typename Kernel_traits::Gmem_tile_mask;
+    // conctructor
+    Gmem_tile_mask gmem_mask(params, binfo, tidx, loop_step_idx);
+
+    // Allocate the global memory tile loader for bias.
+    using Gmem_tile_bias = typename Kernel_traits::Gmem_tile_bias;
+    // conctructor
+    Gmem_tile_bias gmem_bias(params, binfo, tidx, loop_step_idx);
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h:161]: move these inits into has_attn or has_bias?

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h [hunk 6]
```diff
@@ -31,7 +31,7 @@ inline __device__ void dot_do_o(const uint4 (&do_)[M], const uint4 (&o)[M], cons
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
-template<typename Kernel_traits, bool Is_dropout, bool Is_causal, bool Is_first, bool Is_last, typename Params, typename Prng>
+template<typename Kernel_traits, bool Is_dropout, bool Is_causal, bool has_attn, bool has_bias, bool Is_first, bool Is_last, typename Params, typename Prng>
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h]: rename all `has_attn` to `has_attn_mask`. and `has_bias` to `has_attn_bias` ?

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h (hunk 4 above)
Comment (guolinke) [reply csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h]: also check these conditions more carefully.

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h [hunk 7]
```diff
@@ -148,6 +148,20 @@ inline __device__ void compute_dq_dk_dv_1xN_one_iter(const Params &params, Prng
...
+    // conctructor
+    Gmem_tile_mask gmem_mask(params, binfo, tidx, loop_step_idx);
+
+    // Allocate the global memory tile loader for bias.
+    using Gmem_tile_bias = typename Kernel_traits::Gmem_tile_bias;
+    // conctructor
+    Gmem_tile_bias gmem_bias(params, binfo, tidx, loop_step_idx);
+
+    using Gmem_tile_ds = typename Kernel_traits::Gmem_tile_ds;
+    Gmem_tile_ds gmem_ds(params, binfo, tidx, loop_step_idx);
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h:162]: ds should be disable during forward?

File: csrc/flash_attn/src/fmha_fprop_kernel_1xN.h (hunk 2 above)
Comment (robotcator) [reply csrc/flash_attn/src/fmha_fprop_kernel_1xN.h]: done

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h (hunk 4 above)
Comment (robotcator) [reply csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h]: done.

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h (hunk 5 above)
Comment (robotcator) [reply csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h:161]: this part is used for `declaration`.

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h (hunk 6 above)
Comment (robotcator) [reply csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h]: done

File: csrc/flash_attn/fmha_api.cpp (hunk 1 above)
Comment (robotcator) [reply csrc/flash_attn/fmha_api.cpp:484]: we didn't use the block reduce method to get the `dbias`, so we store the ds and start a new small kernel to compute `dbias`.

File: csrc/flash_attn/src/fmha_fprop_kernel_1xN.h [hunk 8]
```diff
@@ -52,6 +52,7 @@ struct Gemm_Q_K_base {
     using Mma_tile_p = fmha::Hmma_tile<Cta_tile_p>;
 
     static constexpr int SMEM_BYTES_SOFTMAX = Cta_tile_p::M * Cta_tile_p::WARPS_N * sizeof(float) * 2;
+    // ?
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_fprop_kernel_1xN.h]: remove these comments

File: csrc/flash_attn/src/fmha_fprop_kernel_1xN.h [hunk 9]
```diff
@@ -255,13 +257,26 @@ inline __device__ void device_1xN_(const Params &params, const int bidb, const i
     if( binfo.stop_early(loop_step_idx * Cta_tile_p::N) ) return;
 
     Gemm1 gemm_q_k(smem_, tidx);
+
     // Allocate the global memory tile loader for Q.
     Gmem_tile_q gmem_q(params.q_ptr, params.q_row_stride_in_elts, params.q_head_stride_in_elts, binfo, tidx, true);
+
```
Comment (guolinke) [inline csrc/flash_attn/src/fmha_fprop_kernel_1xN.h]: remove empty line changes

File: csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h (hunk 7 above)
Comment (robotcator) [reply csrc/flash_attn/src/fmha_dgrad_kernel_1xN_loop.h:162]: yes, this part is for backward.

**Review by guolinke — APPROVED** [review]

**Pull request merged**